### DISCUSSION TITLE
Github basic auth deprecation

### DISCUSSION THREAD

**Discussion by You-Honey** [renovatebot/renovate#10560] [Request Help]
**How are you running Renovate?**

- [x] WhiteSource Renovate hosted app on github.com
- [ ] Self hosted

If using the hosted app, please skip to the next section.
Otherwise, if self-hosted, please complete the following:

Please select which platform you are using:

- [ ] Azure DevOps (dev.azure.com)
- [ ] Azure DevOps Server
- [ ] Bitbucket Cloud (bitbucket.org)
- [ ] Bitbucket Server
- [ ] Gitea
- [x] github.com
- [ ] GitHub Enterprise Server
- [ ] gitlab.com
- [ ] GitLab self-hosted

Renovate version: ...

Here's my config:

```
 {
  "enabled": true,
  "enabledManagers": ["composer"],
  "packageRules": [
    {
      "excludePackagePatterns": ["MYORG/MYPACKAGE"],
      "enabled": false,
      "matchPackagePatterns": ["*"]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "encrypted": {
        "username": "REMOVED",
        "password": "REMOVED"
      }
    }
  ]
}
```

Which work fine but I just recieved this email:

```
Hi @REMOVED,

You recently used a password to access the repository at MYORG/MYANOTHERPACKAGE with git using git/2.32.0.

Basic authentication using a password to Git is deprecated and will soon no longer work. Visit https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/ for more information around suggested workarounds and removal dates.
```

How can I get composer to access private repositories on github after the deprecation of basic authentication?
...

**Relevant debug logs**

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

...

**Comment by herndlm** [comment]
I think it tells you to use tokens instead of username/password, right? Renovate is supporting tokens too, see https://docs.renovatebot.com/private-modules/#npmjs-registry-token-example
So I guess you can create a personal token and configure that instead of username/password.

**Reply by You-Honey** [reply]
@monojp Thank you for the quick answer.

I tried it like this:

```
 {
  "enabled": true,
  "enabledManagers": ["composer"],
  "packageRules": [
    {
      "excludePackagePatterns": ["MYORG/MYPACKAGE"],
      "enabled": false,
      "matchPackagePatterns": ["*"]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "encrypted": {
        "token": "REMOVED"
      }
    }
  ]
}
```

But received:

 ```
  Failed to execute git clone --mirror -- 'https://**redacted**:***@github.com/MYORG/MYPACKAGE.git' '/tmp/renovate-cache/others/composer/vcs/git-github.com-MYORG-MYPACKAGE.git/'  
                                                                                                                                                                                                                                                   
  Cloning into bare repository '/tmp/renovate-cache/others/composer/vcs/git-github.com-MYORG-MYPACKAGE.git'...                                                                                                                   
  remote: Invalid username or password.                                                                                                                                                                                                            
  fatal: Authentication failed for 'https://github.com/MYORG/MYPACKAGE.git/' 
  ```
  
   Did I enter the token wrong?

**Comment by rarkins** [comment]
What's the original reason for adding the host rule?

**Reply by You-Honey** [reply]
Renovate was able to edit the `composer.json` file but was not able to generate `composer.lock` file as there was private repositories in `composer.json`.

**Reply by viceice** [reply]
You need to add removate app to your private repo too, otherwise the hosted app is not able to create a token for both repos.

So both repos need to be public or private.

**Reply by viceice** [reply]
With hosted renovate it seems not possible to use a custom github.com token.

Did you verified your token can clone that repo?

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/blob/fbbbce608e5df637ac5a6447774d38e03b2e79fc/lib/manager/composer/artifacts.ts#L34-L42

**Reply by You-Honey** [reply]
I'm not using hosted renovation. I'm using the app on github's marketplace.

These are the privileges on the personal access token:

```
 repo Full control of private repositories
 repo:status Access commit status
 repo_deployment Access deployment status
 public_repo Access public repositories
 repo:invite Access repository invitations
 security_events Read and write security events
 ```

**Reply by rarkins** [reply]
Hosted renovate = the app. Is the app installed into all repos in this thread?

**Reply by You-Honey** [reply]
@rarkins Hey, I added renovate to the other repository but I still got this:

```
  [RuntimeException]                                                                                                                                                                                                                               
  Failed to execute git clone --mirror -- 'https://**redacted**:***@github.com/MYORG/MYOTHERREPOSITORY.git' '/tmp/renovate-cache/others/composer/vcs/git-github.com-MYORG-MYOTHERREPOSITORY.git/'  
                                                                                                                                                                                                                                                   
  Cloning into bare repository '/tmp/renovate-cache/others/composer/vcs/git-github.com-MYORG-MYOTHERREPOSITORY.git'...                                                                                                                   
  remote: Invalid username or password.                                                                                                                                                                                                            
  fatal: Authentication failed for 'https://github.com/MYORG/MYOTHERREPOSITORY.git/'
  ```
  
☝️ And as stated previously renovate was activated on https://github.com/MYORG/MYOTHERREPOSITORY.

I have only been able to get composer work with private repositories by using username and password but they will be deprecated in a week or so.

How can I get ssh / oauth or token to work with composer and renovate?

**Reply by rarkins** [reply]
OK so prerequisites are:
- Both repos are added to Renovate
- Both repos are private

Those are met now, right?

The the logic for building the composer auth is here: https://github.com/renovatebot/renovate/blob/11f89147acfa2d8339991bb03a5d88ced214d8b6/lib/manager/composer/artifacts.ts#L30-L69

There *should* be a line for `github.com` as a result of that.

**Reply by You-Honey** [reply]
I even tried the `COMPOSER_AUTH` way like this:

```
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": "{\"github-oauth\": {\"github.com\": \"my-personal-token\"}}"
    }
```

Did I misunderstand the docs?

**Reply by rarkins** [reply]
You misunderstood, token is used as a string to go into http headers

**Reply by You-Honey** [reply]
It would help me a lot if you showed me how I add the token to the http headers.

**Reply by rarkins** [reply]
You add a string to the "token" field in hostRules and the header insertion is already done for you. Every hostRules example with a token is like this

**Reply by You-Honey** [reply]
I'm not 100% sure if I follow you.

Are you saying I should change this:

```
 {
  "enabled": true,
  "enabledManagers": ["composer"],
  "packageRules": [
    {
      "excludePackagePatterns": ["johnpbloch/wordpress"],
      "enabled": false,
      "matchPackagePatterns": ["*"]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": "{\"github-oauth\": {\"github.com\": \"MY_TOKEN\"}}"
    }
  ]
}
```

To this:

```
 {
  "enabled": true,
  "enabledManagers": ["composer"],
  "packageRules": [
    {
      "excludePackagePatterns": ["johnpbloch/wordpress"],
      "enabled": false,
      "matchPackagePatterns": ["*"]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": "MY_TOKEN"
    }
  ]
}
```

**Reply by rarkins** [reply]
Yes for the token field, but matchHost=github.com might break the app if you override the app's own token

**Reply by You-Honey** [reply]
What matchHost should I use? All private composer packages are located in github.

I changed the config to the one I listed on the previous message:

```
 {
  "enabled": true,
  "enabledManagers": ["composer"],
  "packageRules": [
    {
      "excludePackagePatterns": ["johnpbloch/wordpress"],
      "enabled": false,
      "matchPackagePatterns": ["*"]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": "MY_TOKEN"
    }
  ]
}
```

And I still received this:

```
  [RuntimeException]                                                                                                                                                                                                                               
  Failed to execute git clone --mirror -- 'https://**redacted**:***@github.com/our-company/our-package.git' '/tmp/renovate-cache/others/composer/vcs/git-github.com-our-company-our-package.git/'  
                                                                                                                                                                                                                                                   
  Cloning into bare repository '/tmp/renovate-cache/others/composer/vcs/git-github.com-our-company-our-package.git'...                                                                                                                   
  remote: Invalid username or password.                                                                                                                                                                                                            
  fatal: Authentication failed for 'https://github.com/our-company/our-package.git/'    
  ```
  
   The token is a personal access token created on my Github account and I have admini privileges to the `our-package` repository.

**Reply by rarkins** [reply]
I'm not sure you should need to add a host rule at all for this, if all repos are added to the app. 

Instead, it needs to be debugged about what credentials are added to the auth.json and why that's not working as is

**Reply by You-Honey** [reply]
I'm not sure how to debug it. I tried viewing the log from renovate dashboard but there's no mention of auth.json.

Could you provide me clear instructions how can I get Personal Access Token to work with renovate for composer packages?

The docs mention composer basically with a "one liner":

```
composer
Any hostRules token for github.com or gitlab.com are found and written out to COMPOSER_AUTH in env for Composer to parse. Any hostRules with hostType=packagist are also included.
```

An example config would greatly help many of us.

**Reply by You-Honey** [reply]
Hey @rarkins I tried these now:

```
{
  "enabled": true,
  "enabledManagers": [
    "composer"
  ],
  "packageRules": [
    {
      "excludePackagePatterns": [
        "johnpbloch/wordpress"
      ],
      "enabled": false,
      "matchPackagePatterns": [
        "*"
      ]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": "{\"github-oauth\":{\"github.com\":\"MY_TOKEN\"}}"
    }
  ]
}
```

and 

```
{
  "enabled": true,
  "enabledManagers": [
    "composer"
  ],
  "packageRules": [
    {
      "excludePackagePatterns": [
        "johnpbloch/wordpress"
      ],
      "enabled": false,
      "matchPackagePatterns": [
        "*"
      ]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": {
        "github-oauth": {
          "github.com": "MY_TOKEN"
        }
      }
    }
  ]
}
```

and

```
{
  "enabled": true,
  "enabledManagers": [
    "composer"
  ],
  "packageRules": [
    {
      "excludePackagePatterns": [
        "johnpbloch/wordpress"
      ],
      "enabled": false,
      "matchPackagePatterns": [
        "*"
      ]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": "COMPOSER_AUTH={ \"github-oauth\": { \"github.com\": \"MY_TOKEN\" } }"
    }
  ]
}
```

None of them gave access to our private repositories, located in github.

As a test I copied the exact same `composer.json` file inside this container:
`docker run --rm --interactive --env COMPOSER_AUTH='{ "github-oauth": { "github.com": "MY_TOKEN" } }' --tty --volume $PWD:/app composer bash`

And `composer install` run without any problems. Access to our private repositories was granted.

Could you give me an example what this means as I am not able to get it to work:

```
Any hostRules token for github.com or gitlab.com are found and written out to COMPOSER_AUTH in env for Composer to parse. Any hostRules with hostType=packagist are also included.
```

**Reply by viceice** [reply]
@JuhaniGeniem The hosted renovate app already includes a github token which is used here:
https://github.com/renovatebot/renovate/blob/3be44469bb98600a8f839887f9cb3c5c6bf975f3/lib/manager/composer/artifacts.ts#L33-L41

So you shouldn't add that anyways.

You need to make sure the hosted app is added to all your private composer repos on github, then the app generated token should have access to all private repos.

BTW: Are all repos private? Or do you try to update private composer deps in a public project?

**Reply by You-Honey** [reply]
My project is private and some of the dependencies are public and some are private.

So far I've only added the project and one of the private dependencies to Renovate. I have not added all dependencies to Renovate yet. I'm getting the error:

```
  [RuntimeException]                                                                                                                                                                                                                               
  Failed to execute git clone --mirror -- 'https://**redacted**:***@github.com/MYORG/MYOTHERREPOSITORY.git' '/tmp/renovate-cache/others/composer/vcs/git-github.com-MYORG-MYOTHERREPOSITORY.git/'  
                                                                                                                                                                                                                                                   
  Cloning into bare repository '/tmp/renovate-cache/others/composer/vcs/git-github.com-MYORG-MYOTHERREPOSITORY.git'...                                                                                                                   
  remote: Invalid username or password.                                                                                                                                                                                                            
  fatal: Authentication failed for 'https://github.com/MYORG/MYOTHERREPOSITORY.git/'
  ```
  
  On the private dependency I have added to Renovate. Which makes me think Renovate's composer does not have access to the private repository even though I've added it to Renovate.

**Reply by herndlm** [reply]
If composer generates a lockfile in general it needs to have access to all deps, even those it doesn't upgrade in that PR. So in your case all private repos need to be accessible then too. This is some Composer internal stuff that I also learned the hard way :)

**Reply by You-Honey** [reply]
Interesting. I'll check it out tomorrow and let you know.

Thank you!

**Reply by You-Honey** [reply]
@monojp @viceice @rarkins 

I have now simplified this experiment a lot. I now have 3 repositories:

- devgeniem/renovate-test, the repository that I want to get updated by Renovate
- devgeniem/devops-renovate-config, the repository that holds Renovate configs that are extended in renovate-test
- devgeniem/renovate-private-repository-test, a dummy private repository

My Renovate config looks like this:
**devgeniem/renovate-test/.github/renovate.json**:
```
{
  "extends": ["github>devgeniem/devops-renovate-config:default", "github>devgeniem/devops-renovate-config:wordpress"]
}
```

**devgeniem/devops-renovate-config/default.json**:
```
{
  "enabled": false
}
```

**devgeniem/devops-renovate-config/wordpress.json**:
```
{
  "enabled": true,
  "enabledManagers": [
    "composer"
  ],
  "packageRules": [
    {
      "excludePackagePatterns": [
        "johnpbloch/wordpress"
      ],
      "enabled": false,
      "matchPackagePatterns": [
        "*"
      ]
    }
  ]
}
```

And my **devgeniem/renovate-test/composer.json** looks like this:
```
{
    "name": "devgeniem/renovate-test",
    "type": "project",
    "license": "MIT",
    "config": {
        "preferred-install": "dist"
    },
    "repositories": {
        "devgeniem/renovate-private-repository-test": {
            "type": "git",
            "url": "git@github.com:devgeniem/renovate-private-repository-test.git"
        }
    },
    "require": {
        "devgeniem/renovate-private-repository-test": "^1.0",
        "johnpbloch/wordpress": "5.4.2"
    }
}
```

 And my **devgeniem/renovate-test/composer.json** looks like this:
 ```
{
    "name": "devgeniem/renovate-private-repository-test",
    "require": {}
}
```

I have added three repositories to Renovate:
<img width="503" alt="Screenshot 2021-06-30 at 9 02 55" src="https://user-images.githubusercontent.com/22659221/123909684-06eb6c80-d982-11eb-905f-19bdc87578ef.png">

And still I receive the same error:
<img width="859" alt="Screenshot 2021-06-30 at 9 03 50" src="https://user-images.githubusercontent.com/22659221/123909744-208cb400-d982-11eb-9873-c8d1561cf546.png">

Am I still doing something wrong or can we determine that Renovate does not grant access of a private repository to composer even if said repository has been added to Renovate?

**Reply by rarkins** [reply]
based on the screenshot it appears that Composer is using a credential to try to git clone, but it's being treated as an OAuth token and having `:x-oath-basic` appended, which won't work. It would seem like we need to provide Composer with the token in a different field maybe.

**Reply by You-Honey** [reply]
@rarkins How do we proceed?

I get this error:
```
[RuntimeException]                                                                                                                                                                                                                                                       
Failed to execute git clone --mirror -- 'https://**redacted**:x-oauth-basic@github.com/devgeniem/my-private-repository.git' '/tmp/renovate-cache/others/composer/vcs/git-github.com-devgeniem-my-private-repository.git/'  
```

If I grab the git clone from the error message and run it manually while replacing the **redacted** with my key:

```
git clone --mirror -- 'https://ghp_mykey_mykey_mykey:x-oauth-basic@github.com/devgeniem/my-private-repository.git' '/tmp/renovate-cache/others/composer/vcs/git-github.com-devgeniem-my-private-repository.git/'  
```

It works fine. I have a feeling the my ghp key is not being used correctly on the **redacted** spot.

I've tried:

```
{
  "enabled": true,
  "enabledManagers": [
    "composer"
  ],
  "updateLockFiles": true,
  "packageRules": [
    {
      "excludePackagePatterns": [
        "johnpbloch/wordpress"
      ],
      "enabled": false,
      "matchPackagePatterns": [
        "*"
      ]
    }
  ],
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": {"http-basic": {"github.com": { "username": "ghp_mykey_mykey_mykey", "password": "" }}}
    }
  ]
}
```

And:
```
...
"token": {"http-basic": {"github.com": { "username": "ghp_mykey_mykey_mykey", "password": "ghp_mykey_mykey_mykey" }}}
...
```

And:
```
...
"token": {"http-basic": {"github.com": { "username": "JuhaniGeniem", "password": "ghp_mykey_mykey_mykey" }}}
...
```

Am I writing the ghp key to wrong place?

**Reply by rarkins** [reply]
Within `hostRules`, `token` should be a string, not object

**Reply by You-Honey** [reply]
@rarkins I changed it to this:

```
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": "{\"http-basic\": {\"github.com\": { \"username\": \"JuhaniGeniem\", \"password\": \"ghp_mykey\" }}}"
    }
  ]
  ```

But I received same errors. Should I try some other string?

Oh and a huge thank you for helping me. I appreciate it.

**Reply by rarkins** [reply]
I don't think anything is going to work without some code changes. First step is to work out exactly what the `auth.json` needs to look like:
- If we use the app token only, or
- If you provision a host rule manually with a ghp_ token

**Reply by You-Honey** [reply]
@rarkins 

Renovate's docs says:

```
Any hostRules token for github.com or gitlab.com are found and written out to COMPOSER_AUTH in env for Composer to parse. Any hostRules with hostType=packagist are also included.
```

As far as I understand it means that everything written under `token` should be printed to `COMPOSER_AUTH` as is.

I created a test container where I had PHP + Composer and I executed this script:

```
#!/bin/bash

export COMPOSER_AUTH='{"http-basic": {"github.com": { "username": "JuhaniGeniem", "password": "ghp_mykey" }}}'

composer install --ignore-platform-reqs
```

And it worked beautifully. It was able to download all private repositories.

Shouldn't this:

```
  "hostRules": [
    {
      "matchHost": "github.com",
      "token": "{\"http-basic\": {\"github.com\": { \"username\": \"JuhaniGeniem\", \"password\": \"ghp_mykey\" }}}"
    }
  ]
```
  
Do the exact same thing? What I don't understand is why Renovate runs this git command:

```
git clone --mirror -- 'https://**redacted**:x-oauth-basic@github.com/devgeniem/my-private-repository.git'
```

When it should run this:

```
git clone --mirror -- 'https://JuhaniGeniem:**redacted**@github.com/devgeniem/my-private-repository.git'
```

Or am I missing something here?

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/blob/3505508c9c03599eb64cbd9792018a18de4db42d/lib/manager/composer/artifacts.ts#L35-L74

☝️
